pkg/cloudprovider/aws/instance.go

```diff
@@ -105,6 +128,37 @@ func (p *InstanceProvider) Terminate(ctx context.Context, node *v1.Node) error {
 	return nil
 }
 
+// The intention is to remove Offerings from the provided possible InstanceTypes based on recently observed ICEs, which will
+// indirectly impact the packer to keep it from spinning its wheels on packings that are doomed to fail due to EC2 capacity constraints.
+func (p *InstanceProvider) DiscardICEdInstanceTypes(ctx context.Context, instanceTypes []InstanceType) []cloudprovider.InstanceType {
+	var cleanedInstanceTypes = []cloudprovider.InstanceType{}
+	for index, instanceType := range instanceTypes {
+		odIceZones, hasODIces := p.iceCache.Get(createIceCacheKey(instanceType.Name(), v1alpha1.CapacityTypeOnDemand))
+		spotIceZones, hasSpotIces := p.iceCache.Get(createIceCacheKey(instanceType.Name(), v1alpha1.CapacityTypeSpot))
+		if !hasODIces && !hasSpotIces {
+			cleanedInstanceTypes = append(cleanedInstanceTypes, &instanceTypes[index])
+			continue
+		}
+		cleanedOfferings := []cloudprovider.Offering{}
+		for _, offering := range instanceType.Offerings() {
+			if hasODIces && offering.CapacityType == v1alpha1.CapacityTypeOnDemand {
+				if !odIceZones.(sets.String).Has(offering.Zone) {
+					cleanedOfferings = append(cleanedOfferings, offering)
+				}
+			} else if hasSpotIces {
+				if !spotIceZones.(sets.String).Has(offering.Zone) {
+					cleanedOfferings = append(cleanedOfferings, offering)
+				}
+			}
+		}
+		if len(cleanedOfferings) > 0 {
+			instanceType.AvailableOfferings = cleanedOfferings
+			cleanedInstanceTypes = append(cleanedInstanceTypes, &instanceTypes[index])
```

Comment:
This syntax seemed funky to me but doing I get a linting error if I just append `*instanceType` - there's certainly some Go idiosyncrasy I don't follow here.

Reply:
I believe this is the issue I mentioned above where the range variable in a for loop uses one memory address and overrides it for each iteration. So you'd need to set another temp variable in the loop, and take the address of it when appending.